src/db.c

```diff
@@ -1502,31 +1490,34 @@ int keyIsExpired(redisDb *db, robj *key) {
  * is via lookupKey*() family of functions.
  *
  * The behavior of the function depends on the replication role of the
- * instance, because slave instances do not expire keys, they wait
- * for DELs from the master for consistency matters. However even
- * slaves will try to have a coherent return value for the function,
- * so that read commands executed in the slave side will be able to
- * behave like if the key is expired even if still present (because the
- * master has yet to propagate the DEL).
+ * instance, because replicas do not expire keys. They wait for DELs
+ * from the master for consistency matters. The exception is write
+ * operations to writable replicas. However, even read-only replicas
+ * will try to have a coherent return value of the function, so that
+ * read commands executed on the replica will be able to behave like if
+ * the key is expired even if still present (because the master has yet
+ * to propagate the DEL).
  *
  * In masters as a side effect of finding a key which is expired, such
  * key will be evicted from the database. Also this may trigger the
  * propagation of a DEL/UNLINK command in AOF / replication stream.
  *
  * The return value of the function is 0 if the key is still valid,
  * otherwise the function returns 1 if the key is expired. */
-int expireIfNeeded(redisDb *db, robj *key) {
+int expireIfNeeded(redisDb *db, robj *key, int expire_on_replica) {
```

Comment:
Do we want to add an assert here?

```suggestion
int expireIfNeeded(redisDb *db, robj *key, int expire_on_replica) {
    assert(!expire_on_replica || server.masterhost == NULL || !server.repl_slave_ro);
```

Reply:
so this assertion will fire if someone calls lookupKeyWrite in a read command when that command runs on a read-only replica, right?
i'd like to catch these cases, but since we don't have much tests running on a writable replica, i don't think it'll catch them.
instead how about just asserting that `c->cmd->flags` has `CMD_WRITE`? (with a proper comment)
this will crash a lot sooner if we have such a bug.

Reply:
I tried adding an assert and it catches some stuff already:

* Current command can be EXEC, DEBUG LOADAOF, EVAL or EVALSHA, which don't have the CMD_WRITE flag. It works if we check for these too (four more lines), or we just check for the CMD_READONLY flag instead. I did the latter.

* lookupKeyWrite is called from handleClientsBlockedOnKeys. Here, we don't know if the key is blocked by a read or a write command. XREAD is readonly for example, so if that's called on a replica, we have a problem. I'm changing this to lookupKeyRead.

* PFCOUNT does lookupKeyWrite, but it's a readonly command. I'm changing it to lookupKeyRead.

Reply:
Considering I had to fix some test modules which marked themselves as "readonly" when they were not, we may break some other modules with this PR. Something to highlight for the release notes.

Reply:
> Current command can be EXEC, DEBUG LOADAOF, EVAL or EVALSHA, which don't have the CMD_WRITE flag. It works if we check for these too (four more lines), or we just check for the CMD_READONLY flag instead. I did the latter.

did you use `c->cmd` or `c->last_cmd`?
i don't mind using the READONLY flag, but in theory commands that access the keyspace, should have one or the other, not both or none.
the commands that'll have none are the ones that you mentioned, as well as PING / CLIENT etc, i.e. not touching the keyspace, so i think we better resolve this than switch to the other flag. 

> lookupKeyWrite is called from handleClientsBlockedOnKeys. Here, we don't know if the key is blocked by a read or a write command. XREAD is readonly for example, so if that's called on a replica, we have a problem. I'm changing this to lookupKeyRead.

I'd like to resolve this too (i.e. i want XREAD to use read, and BLPOP to use write).
maybe we can somehow check which clients are waiting for that key, and look at `c->last_cmd` and decide which lookup to use according to the right flag? i.e. if at least one is a write, we'll use write?

> PFCOUNT does lookupKeyWrite, but it's a readonly command. I'm changing it to lookupKeyRead.

PFCOUNT is an odd command, it's marked as read-only but it does modify the key and replicates itself.
i'm not yet sure what to do, but let's think of the consequences and not dismiss that just yet.

Reply:
I used `c->cmd`. What's the difference?

For the blocking commands, we can use lookupKeyRead with nostats flags to check if the key exists check its type, and delegate to the handler per type. Then, there can be another lookupKey depending on command. OK?

Reply:
> Considering I had to fix some test modules which marked themselves as "readonly" when they were not, we may break some other modules with this PR. Something to highlight for the release notes.

that may mean that we can't afford to add that assertion. we need to be ABI compatible with old modules. maybe exclude module commands from that assertion?

Reply:
> I used c->cmd. What's the difference?

IIRC `c->cmd` will change inside EXEC to point to the actual command that's running, and `last_cmd` will remain pointing to EXEC.
`realcmd` (in `call`) will be like `c->cmd` but is not modified when the command gets re-written (like GEOADD being redirected to ZADD).

i think `c->cmd` should have worked correctly for you, please check again.

> For the blocking commands, we can use lookupKeyRead with nostats flags to check if the key exists check its type, and delegate to the handler per type. Then, there can be another lookupKey depending on command. OK?

i don't like to do double lookup (efficiency).. maybe we should add another method for just adding stats and call it in these blocked commands processing?

Reply:
OK, thx. So, I can confirm that during exec, `server.current_client->cmd` is the queued command, but when handleClientsBlockedOnKeys is called afterwards, `server.current_client->cmd` is exec.

Although the command function (e.g. blockingGenericZpopCommand) is passed a client, lookupKey and expireIfNeeded aren't. They only have access to `server.current_client`, which is not the same client in this case. A test case which crashes is `"MULTI/EXEC is isolated from the point of view of $pop"` in unit/zset.tcl.

For eval, the situation is very similar to exec: The current client's cmd is set to eval while the unblocked client is running the BLPOP command (or similar).

One idea is to set `server.current_client` to the client which is being served (running ZPOPMIN, BLMOVE, etc.) so that the assert in expireIfNeeded works, and change it back to the old current_client (the one which triggered the unblocking using exec or eval) afterwards. It feels like a hack though.

For debug loadaof, similarily, maybe it's possible to temporarily set `server.current_client` to the fake client which executes commands from the AOF file.

Setting `current_client` seem like a hack, right? Passing the client around to lookupKey and expireIfNeeded might be a better idea but it's a lot of work to do it everywhere. Another option is to just scrap the assert...

Reply:
> OK, thx. So, I can confirm that during exec, server.current_client->cmd is the queued command, but when handleClientsBlockedOnKeys is called afterwards, server.current_client->cmd is exec.

ok, i didn't realize that we're discussing blocked clients here too... but then i don't understand how this is related to exec (blocking commands aren't working in a transaction.

> Although the command function (e.g. blockingGenericZpopCommand) is passed a client, lookupKey and expireIfNeeded aren't. They only have access to server.current_client, which is not the same client in this case. A test case which crashes is "MULTI/EXEC is isolated from the point of view of $pop" in unit/zset.tcl.

correct me if i'm wrong, but it's not just the fact the current_client is not set, and the client is not passed to lookupKey. the problem is that the lookupKey is done before looping on the list of clients that are blocked on that key (so it's one lookup possibly for serving many clients and many different commands). in that case the only way out IMHO is to tell lookupKey not to record stats, and instead have the individual blocked command handlers update the stats later.

> One idea is to set server.current_client to the client which is being served (running ZPOPMIN, BLMOVE, etc.) so that the assert in expireIfNeeded works, and change it back to the old current_client (the one which triggered the unblocking using exec or eval) afterwards. It feels like a hack though.

please note that in unstable this is already handled (#9422), i think you should update your branch.

> For debug loadaof, similarily, maybe it's possible to temporarily set server.current_client to the fake client which executes commands from the AOF file.

seems right, please go ahead.

> Setting current_client seem like a hack, right? Passing the client around to lookupKey and expireIfNeeded might be a better idea but it's a lot of work to do it everywhere. Another option is to just scrap the assert...

yes, the use of globals in redis may be a problem some day, but let's not deal with it now.
i think this assert may be very useful, if we write any code that relies on these flags, this assert will give us an indication that we're not introducing bugs.

Reply:
> but then i don't understand how this is related to exec (blocking commands aren't working in a transaction.

The blocked client (e.g. BLPOP) is not in a transaction. It's another client which unblocks the blocked client by e.g. RPUSH in a transaction (exec) or Lua (eval).

> in unstable this [setting server.current_client] is already handled

Nice! I didn't see that. I will update and then do the same for debug loadaof.

(I was just starting to implement a lookupKeyForClient with an extra client parameter, only for use when the client is not the current client. The old lookup functions would remain and use server.current_client as a default. But I didn't get that far.)

Reply:
Done. See updated top comment and commit messages. The assertion now looks like this:

```C
    if (expire_on_replica) {
        client *c = server.in_eval ? server.lua_client : server.current_client;
        serverAssert(!c || !c->cmd || (c->cmd->flags & CMD_WRITE));
    }
```

Now I can see two risks:

1. Modules will break (module write commands without the write flag)
2. This scenario:
    * (client 1) `blpop k`
    * (client 2) `eval 'multi; rpush k 42; exec'` -> when blpop is unblocked, `server.in_eval` is 1 and the lua_client's current command is exec, so the assertion might fail. (I'll need a test case to be sure.)

(2) can be mitigated by just modifying the assert a little, but (1) is intentional. I think (1) needs to be in the release notes. I had to update many test modules and it may very well happen to real modules out there.

Reply:
I think we have to exempt modules from that assertion.
the new redis release needs to be compatible with old modules, we need to consider the implications of modules using the wrong lookup.

The main difference in REDISMODULE_WRITE, is that RM_OpenKey returns a non-null RedisModuleKey even if the key is not found in redis, in which case `kp->value` will be NULL.
And modules that do that, expect that and could still use RM_ModuleTypeSetValue and RM_ModuleTypeReplaceValue, and unlike redis (e.g. incrDecrCommand) don't attempt to access the robj directly without checks.

So let's try to consider the implications of our change (assuming the assertion is skipped).
We moved the expiration attempt from lookupKeyRead to lookupKey, so it now affects REDISMODULE_WRITE too.
And the difference is only when used on a replica (readonly or writable, since modules could use that on RO commands), when a command is executed from the non-master client.
And the implication would be that both before and after our fix, RM_OpenKey succeeds, however after our fix, any attempt to access the value RM_HashGet will fail (only if the key already expired), and in the past it would have succeeded.
Either way, with or without our fix, RM_ModuleTypeSetValue and RM_ModuleTypeReplaceValue will succeed and do the same thing.

anything i'm missing?

regarding BLPOP, can you please describe the flow in the code itself?
AFAICT `handleClientsBlockedOnKeys` is only called by `processCommand` so would only be called when EVAL returns.

Reply:
Makes sense. I've added CMD_MODULE to the assert, meaning that all module commands can use lookupKeyWrite. Technically, we don't need the commit adding the write flag to all the test modules "Fixup: Modules test commands marked as readonly" 068e7a1 and "Add WRITE flag to all test module write commands" 32a7045. Should I revert them?

> regarding BLPOP, can you please describe the flow in the code itself?

I was wrong here. Multi and exec can't be used from within Lua scripts, so this can't happen.

Btw, I'm sorry for the rebase and force-push. It was sort of automatic from my side, since many other projects forbid merge commits. It does make sense if you always make squash commits though. I'll do merge from now on.